### PR TITLE
add validation for static pods to have a name

### LINKED ISSUE #119135: kubelet applies defaults to Pods and causes could not process manifest file
### What happened?

When I put a Pod yaml file without `metadata.name` in `/etc/kubernetes/manifests`, started the kubelet and hoped it would create the corresponding staticPod for me, I got the following error, which also caused the staticPod to not be created.

```
Jul 06 22:46:02 cloud kubelet[2930835]: E0706 22:46:02.905289 2930835 file.go:187] "Could not process manifest file" err="invalid pod: [metadata.name: Invalid value: \"-cloud\": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]" path="/etc/kubernetes/manifests/yurthub.yaml"
```

I then read the kubelet code and quickly found the problem. When kubelet tries to decode the pod yaml file, it will apply default values and validate the pod. The function to apply the defaults is in `pkg/kubelet/config/common.go L56`.

In this function, the `metadata.name` is set.

```go
pod.Name = generatePodName(pod.Name, nodeName)

// Generate a pod name that is unique among nodes by appending the nodeName.
func generatePodName(name string, nodeName types.NodeName) string {
	return fmt.Sprintf("%s-%s", name, strings.ToLower(string(nodeName)))
}
```

This causes an error when `metadata.name` is not set, because it causes the '-' character to be at the front.

I think kubelet should circumvent this error, because `metadata.name` is optional, It is possible for this field not to be set.

### What did you expect to happen?

When Pod's `metadata.name` is not set, the '-' character at the start should be avoided, as it will cause an error.

### How can we reproduce it (as minimally and precisely as possible)?

Prepare a Pod yaml file without `metadata.name` and set it into `/etc/kubernetes/manifests`.

### Anything else we need to know?

If the maintainers think this circumvention change is necessary, I'm happy to contribute code to the kubelet!

### Kubernetes version

<details>

```console
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.17", GitCommit:"a7736eaf34d823d7652415337ac0ad06db9167fc", GitTreeState:"clean", BuildDate:"2022-12-08T11:47:36Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Alibaba Cloud Linux"
VERSION="3 (Soaring Falcon)"
ID="alinux"
ID_LIKE="rhel fedora centos anolis"
VERSION_ID="3"
PLATFORM_ID="platform:al8"
PRETTY_NAME="Alibaba Cloud Linux 3 (Soaring Falcon)"
ANSI_COLOR="0;31"
HOME_URL="https://www.aliyun.com/"
$ uname -a
Linux cloud 5.10.134-14.al8.x86_64 #1 SMP Thu Apr 27 16:46:29 CST 2023 x86_64 x86_64 x86_64 GNU/Linux

</details>


### Install tools

<details>
using kubeadm
</details>


### Container runtime (CRI) and version (if applicable)

<details>

</details>


### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by YTGhost**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Add validation for static Pods to have a name since Pods without names are not working anyways because of this bug.
#### Which issue(s) this PR fixes:

Fixes #119135 

#### Does this PR introduce a user-facing change?

```release-note
Add a check: if a user attempts to create a static pod via the kubelet without specifying a name, they will get a visible validation error.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by YTGhost** [comment]
@SergeyKanzhelev PTAL

File: pkg/kubelet/config/common.go
Comment (bart0sh) [inline pkg/kubelet/config/common.go:73]: I'd suggest to move this code to the top of the function to avoid unnecessary UID generation for the pod without a name.

**Comment by bart0sh** [comment]
/triage accepted
/priority backlog
/ok-to-test
/assign @SergeyKanzhelev 

@YTGhost please provide a release note

**Comment by YTGhost** [comment]
> /triage accepted /priority backlog /ok-to-test /assign @SergeyKanzhelev
> 
> @YTGhost please provide a release note

@bart0sh  release note: If a user attempts to create a static pod via kubelet without specifying a name, they will get a visible validation error.

**Comment by sftim** [comment]
```release-note
Add a check: if a user attempts to create a static pod via the kubelet without specifying a name, they will get a visible validation error.
```
Maybe?

File: pkg/kubelet/config/common.go
Comment (bart0sh) [inline pkg/kubelet/config/common.go:60]: I'd suggest not to mix applying defaults with validation. Do we have another place where static pods are validated? If not, we should probably create it.

File: pkg/kubelet/config/common.go
Comment (bart0sh) [inline pkg/kubelet/config/common.go:126]: There is also tryDecodePodList function below, which is not calling this one.
What about to do this check in the `validation.ValidatePodCreate`? It's called from all places as far as I can see.

Comment (YTGhost) [reply pkg/kubelet/config/common.go:126]: > There is also tryDecodePodList function below, which is not calling this one.

Yeah, sorry I missed that.

> What about to do this check in the `validation.ValidatePodCreate`? It's called from all places as far as I can see.

Since `defaultFn` will be executed before `validation.ValidatePodCreate` and `pod.Name` will be assigned a value in `defaultFn`, I think it still has to be placed before `defaultFn`?

Comment (bart0sh) [reply pkg/kubelet/config/common.go:126]: Correct.
Let's start from adding the same check to the `tryDecodePodList` then.

**Comment by YTGhost** [comment]
@SergeyKanzhelev I think this PR is ready and waiting for your final review, if you could take the time to check it out, it would be appreciated!

**Comment by YTGhost** [comment]
> /lgtm /assign @mrunalp @derekwaynecarr @dchen1107

@bart0sh Hi, I think this PR should be able to be merged now. What is currently holding up the progress?

**Comment by bart0sh** [comment]
@YTGhost 
> @bart0sh Hi, I think this PR should be able to be merged now. What is currently holding up the progress?

The PR is pending an approval. SIG-Node approvers have been assigned to it.

**Review by mrunalp — APPROVED** [review]

**Pull request merged**
